Add symmetrize_from_upper_tri

gregorp90 opened the pull request:
## Summary

Added symmetrize_from_upper_tri.cpp function, which creates a symmetric matrix by using elements in the upper triangular part of the input matrix. Implementation is based on the existing symmetrize_from_lower_tri.cpp.

Changed the second test in symmetrize_from_lower_tri_test.cpp. Before, the input was a symmetric matrix. Now, the input is a non-symmetric square matrix.

## Tests

Added 2 unit tests for symmetrize_from_upper_tri.cpp. First one receives invalid input (non-square matrix) and throws an invalid argument error. The other one receives a non-symmetric square matrix and check if the output is symmetric. The tests are based on symmetrize_from_lower_tri_test.cpp.

## Side Effects

None.

## Release notes

Added symmetrize_from_upper_tri() function.

## Checklist

- [x] Math issue #482

- [x] Copyright holder: Gregor Pirš

rok-cesnovar commented:
@spinkney would you be up for reviewing this one? No rush, whenever you have time.

bbbales2 commented on test/unit/math/prim/fun/symmetrize_from_lower_tri_test.cpp:
```diff
@@ -22,6 +22,6 @@ TEST(Symmetrize, Value) {
   for (int i = 0; i < 4; ++i)
     for (int j = 0; j < 4; ++j)
-      m(i, j) = (i + 1) * (j + 1);
+      m(i, j) = (i + 2) * (j + 1);
```
Ooooh, we should be checking that this function gives the right result.

So build the result:
```
Eigen::MatrixXd m_ref(4, 4);
m_ref << a, b, c, d,
    b, e, f, ...;
EXPECT_MATRIX_EQ(m_sym_ref, symmetrize_from_lower_tri(m));
```

gregorp90 replied:
Thanks Ben, I'll fix these issues.

bbbales2 commented on test/unit/math/prim/fun/symmetrize_from_upper_tri_test.cpp:
```diff
@@ -0,0 +1,27 @@
+  for (int i = 0; i < 4; ++i)
+    for (int j = 0; j < 4; ++j)
+      m(i, j) = (i + 2) * (j + 1);
+  EXPECT_TRUE(is_symmetric(symmetrize_from_upper_tri(m)));
```
Check that this is actually returning the right values too.

bbbales2 reviewed (changes requested):
Even though this function is only in prim, still need to add mixed mode tests (to make sure all the higher order autodiff is working -- it is possible it breaks even if this is fully templated).

Check here: https://github.com/stan-dev/math/blob/develop/test/unit/math/mix/fun/rows_dot_product_test.cpp

But don't do any of the `_matvar` tests cause those are for the varmat types (yeah the naming is a bit funny there lol).

rok-cesnovar commented:
Since this is the first time you are working with the Stan jenkins test, let me give a quick quide.

![image](https://user-images.githubusercontent.com/28476796/107415562-cfd00800-6b13-11eb-8914-2fa5e643aebc.png)

Click "Details" on the failed test.

It seems cpplint has failed

![image](https://user-images.githubusercontent.com/28476796/107415758-fbeb8900-6b13-11eb-9e1d-aa21c2012362.png)

Looking at the complete log here https://jenkins.mc-stan.org/blue/rest/organizations/jenkins/pipelines/Math%20Pipeline/branches/PR-2354/runs/7/nodes/53/steps/56/log/?start=0
we find.
```
Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
```
So you need to add a newline.

You can run cpplint tests localy with

```
mingw32-make cpplint
```

gregorp90 commented:
> Since this is the first time you are working with the Stan jenkins test, let me give a quick quide.
> 
> ![image](https://user-images.githubusercontent.com/28476796/107415562-cfd00800-6b13-11eb-8914-2fa5e643aebc.png)
> 
> Click "Details" on the failed test.
> 
> It seems cpplint has failed
> 
> ![image](https://user-images.githubusercontent.com/28476796/107415758-fbeb8900-6b13-11eb-9e1d-aa21c2012362.png)
> 
> Looking at the complete log here https://jenkins.mc-stan.org/blue/rest/organizations/jenkins/pipelines/Math%20Pipeline/branches/PR-2354/runs/7/nodes/53/steps/56/log/?start=0
> we find.
> 
> ```
> Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
> t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
> Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
> t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
> ```
> 
> So you need to add a newline.
> 
> You can run cpplint tests localy with
> 
> ```
> mingw32-make cpplint
> ```

gregorp90 commented:
Pull request is ready for review. I made the following changes:

1. Added tests to symmetrize_from_lower_tri and symmetrize_from_upper_tri that check if the function is returning the correct values.
2. Added automatic differentiation tests for symmetrize_from_lower_tri and symmetrize_from_upper_tri.

bbbales2 commented:
Looks good to me. I simplified the mix tests a little. Merge if tests pass and you're happy with my changes to the tests.

bbbales2 reviewed (approved):
Good

rok-cesnovar commented:
@bbbales2 Gregor does not have permissions here so you can go ahead and merge yourself unless I beat you to it for that sweet sweet Github points. 

@gregorp90 Thanks! And congrats on your first merged pull request to Stan Math! :tada: :beer:

gregorp90 commented:
@bbbales2, I think the word you were looking for is "somewhat", not "little". Thanks again for reviewing and the comments/simplifications! Really helped in my understanding of the project.

Thanks @rok-cesnovar!
